File: problems/1192_critical_connections_in_a_network/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int timer = 0;
    vector<vector<int>> criticalConnections(int n, vector<vector<int>>& connections) {
        vector<vector<int>> graph(n);
        for (auto& e : connections) {
            graph[e[0]].push_back(e[1]);
            graph[e[1]].push_back(e[0]);
        }
        vector<int> disc(n, -1), low(n, -1);
        vector<vector<int>> result;
        function<void(int, int)> dfs = [&](int u, int parent) {
            disc[u] = low[u] = timer++;
            for (int v : graph[u]) {
                if (v == parent) continue;
                if (disc[v] == -1) {
                    dfs(v, u);
                    low[u] = min(low[u], low[v]);
                    if (low[v] > disc[u]) result.push_back({u, v});
                } else {
                    low[u] = min(low[u], disc[v]);
                }
            }
        };
        dfs(0, -1);
        return result;
    }
};

File: problems/1192_critical_connections_in_a_network/solution.py
class Solution:
    def criticalConnections(self, n: int, connections: list[list[int]]) -> list[list[int]]:
        graph = [[] for _ in range(n)]
        for u, v in connections:
            graph[u].append(v)
            graph[v].append(u)
        disc = [-1] * n
        low = [-1] * n
        result = []
        self.timer = 0
        def dfs(u: int, parent: int) -> None:
            disc[u] = low[u] = self.timer
            self.timer += 1
            for v in graph[u]:
                if v == parent:
                    continue
                if disc[v] == -1:
                    dfs(v, u)
                    low[u] = min(low[u], low[v])
                    if low[v] > disc[u]:
                        result.append([u, v])
                else:
                    low[u] = min(low[u], disc[v])
        dfs(0, -1)
        return result
